### ISSUE TITLE
consider using travis for CI, perhaps with this .travis.yml as a starting point

### ISSUE THREAD

**Issue by thepudds** [dvyukov/go-fuzz#225]
It could be nice to have dvyukov/go-fuzz hooked up to CI for Linux, Mac, and Windows if there is not already something in place.

I am not any type of travis expert, but I put together a basic .travis.yml for dvyukov/go-fuzz here:
https://github.com/thepudds/go-fuzz-with-travis/blob/master/.travis.yml

It is currently running against a fork of go-fuzz, and you can see the results here:
https://travis-ci.org/thepudds/go-fuzz-with-travis

At master, it passes on Linux and Mac while failing on Windows (likely due to #224):
https://travis-ci.org/thepudds/go-fuzz-with-travis/builds/507830409

For testing purposes, I created a branch off of the 2019-03-04 version at commit b73ac4d, which is likely before the problem in #224, and at that revision travis passes on all three platforms (Windows in addition to Linux/Mac):
https://travis-ci.org/thepudds/go-fuzz-with-travis/builds/507830672

From the comment at the top of that .travis.yml:

> There are currently two main sets of tests
> 1. Running go-fuzz-build and go-fuzz on dvyukov/go-fuzz-corpus/png starting with an empty corpus,    and it validates >1 entry appears in the corpus. This uses what was formerly the required arguments like -workdir and -bin.
>  2. Running go-fuzz-build and go-fuzz on dvyukov/go-fuzz/test, which contains some regression tests, etc. This uses the newer ability to not supply any required arguments to go-fuzz-build and go-fuzz.

Obviously, there is more that could be done, but perhaps something along these lines would be useful to start.

I had noticed recently that thepudds/fzgo had started failing in travis on Windows while continuing to pass on Linux/Mac. I suspect it is due to an underlying dvyukov/go-fuzz issue (perhaps #224).

CC @josharian

**Comment by dvyukov** [comment]
This would be totally great!
At the time I did not know about Travis and that I can setup a CI for a project that easily. That's the only reason for not having it.

Please send a PR.
But I have one wish: please move most of the testing logic into some separate location, e.g. `make presubmit` or `presubmit.sh`. This will allow to run it locally as well (and test that the script is syntactically correct at least).

**Comment by josharian** [comment]
Yay! After the windows failure I was also thinking we needed to set up travis. Thanks for doing it! Please do send a PR.

**Comment by thepudds** [comment]
@dvyukov @josharian Great.

Doing something like `presubmit.sh` is a good idea and I would be happy to do it. It would not be a large amount of work, but it might be this weekend or the next before I can get back to it. The current logic is not terribly complicated, but there is a small amount of OS-specific logic (e.g., install `timeout(1)` via coreutils on Windows; do `ln -s /usr/local/bin/gtimeout /usr/local/bin/timeout` on Mac, etc.), and would probably want to start using temporary directories or similar for GOPATH, etc. (instead of assuming a pristine travis environment). At least to start, would you be OK if `presubmit.sh` refused to run on Windows if outside of travis (e.g., if on Windows but $TRAVIS_OS_NAME is not set or some similar type of detection logic)?

Also, what do both of you think think about me sending a PR in its current state, vs. waiting for `presumit.sh`?

I am fine either way.

Finally, I have had some good luck so far with the Windows support with travis, but it is new and still officially an "early release", I think. Overall it was worked well for me, but I think twice I have had the setup of the Windows test environment itself timeout and fail (which made the overall test fail / go red), but in one case it was immediately after I had temporarily disabled travis caching to troubleshoot something, and the other case was on a brand new repo (in other words, no caching in place as far as I understand it). That said, I don't know really know the details of travis caching behavior, or if caching really does explain what I saw, vs. maybe just random bad luck.  In any event, it might make sense to hold off on putting the travis build badge on the go-fuzz readme until there are at least a handful of data points that it seems to work in practice without being too flaky for go-fuzz.

**Comment by dvyukov** [comment]
> Also, what do both of you think think about me sending a PR in its current state, vs. waiting for presumit.sh?

go ahead

> At least to start, would you be OK if presubmit.sh refused to run on Windows if outside of travis (e.g., if on Windows but $TRAVIS_OS_NAME is not set or some similar type of detection logic)?

I would leave the setup part in the travis file for now. And in presubmit.sh assume that the setup is done by user (needs to be done only once so now a large burden, also we don't force people to run presubmit.sh locally).

> In any event, it might make sense to hold off on putting the travis build badge on the go-fuzz readme until there are at least a handful of data points that it seems to work in practice without being too flaky for go-fuzz.

I think the only way is to set it up and see how reliable it is and how many problems it creates.

**Comment by josharian** [comment]
Seems like Travis is working well. @thepudds, any reason not to call this issue fixed and close it?

**Comment by thepudds** [comment]
Hi @josharian, the main thing pending is @dvyukov had suggested in one of the comments above creating a `presubmit.sh` containing most of the testing logic (in order to allow it to be run locally), and I had said I would do that. It is not a lot of work, but just need to find the time to do it ("soon").

This issue could be kept open until that is done, or alternatively it could be closed now given travis does indeed seem to be working.

**Comment by josharian** [comment]
Let's go ahead and close this now that travis is up and running. If/when the pain of not having `presubmit.sh` be factored out gets high enough, someone will do it.
